datafusion/core/tests/optimizer_integration.rs

```diff
@@ -342,7 +343,12 @@ fn test_sql(sql: &str) -> Result<LogicalPlan> {
     let statement = &ast[0];
 
     // create a logical query plan
-    let context_provider = MyContextProvider::default();
+    let now_udf = datetime::functions()
```

Comment:
How much of the optimizer integration test do you think needs Udfs? I am wondering maybe we could just port the tests case for `now()` into core/tests/sql_integration` or something and eave the rest of the optimizer_integration test in the optimizer crate? 

The reason it might be good to leave the optimzer tests in the optimizer crate are 
1. Easier to run `cargo test -p datafusion_optimizer` and run all the relevant tests
2. Ensure that we could still use `datafusion_optimizer` without the function definitions

Reply:
Thinking about this more - the issue is not test related because we can use dev dependencies. The issue is publishing to crates.io because cargo publishes with dev-dependencies and has been for many many years according to the [bug report](https://github.com/rust-lang/cargo/issues/4242). Would it be possible to just fix that by the use of --no-dev-deps  or something similar during publishing?

🤔

Reply:
https://github.com/apache/arrow-datafusion/issues/9579

Reply:
I think moving the tests to the core (what this PR does) is probably the best solution for now, as it correctly reflects the dependencies (the optimizer tests are testing behavior of functions that are not available to optimizer)

Reply:
I moved all but the tests that required now() back to the datafusion/optimizer/tests folder.